Ensure that NVDA  sets Python locale to a valid Windows locale when initializing language.

lukaszgo1 opened the pull request:
### Link to issue number:
Fix-up of #12214
Follow up to #12250
Fixes root cause of #12160
### Summary of the issue:
After NVDA sets its interface language it also attempts to set Python locale to the current language. Before PR #12214  we were just passing the current language code to `locale.setlocale` and catching all possible exceptions. Unfortunately Python `setlocale` is pretty useless on Windows as it tries to normalize the locale name according to the Unix standards (see [Python issue 37945](https://bugs.python.org/issue37945). When we were using wxPython older than 4.1 this was generally not an issue (except for some specific locales under a particular configurations  such as one described in #12160 ) since wxPython was setting locale to a valid one so the problem remained unnoticed. This broke with wxPython 4.1 and the fix was attempted in #12214 but it still was setting Python locale to Unix locales. This is causing two problems:
- We are Windows only so it is important to set locales to something reasonable
- If the locale that is unknown to Windows is set some versions of CRT assumes that the code page of such locale is Utf8 which causes a crash when using `time.localtime` (that was the underlying cause of #12160)

### Description of how this pull request fixes the issue:
Since Python `locale.setlocale` and `locale.getlocale` cannot be trusted on Windows the approach I took avoids all the normalization they offer by creating a valid Windows locale string using [`getLocaleInfoEx`](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getlocaleinfoex) in a form which Python no longer normalizes. This string is then passed to `locale.setlocale`. The new functionality has pretty solid unit test coverage, previous unit tests for `languageHandler` are also updated to account for these changes. As suggested during the review this PR also consolidates all NLS constants we use when retrieving various locale related information from Windows into an enum in `languageHandler`. The  old constants are  deprecated and would no longer be available in 2022.1.
Specific fix for #12160 from #12250 cannot be removed because even though we're setting only complete locales i.e. with a code page and for languages such as Russian or German_Switzerland this crash would no longer occur NVDA supports some locales such as Hindi  for which code page is indeed Utf8 and they would still cause a crash.
### Testing strategy:
Manual testing:
On Windows 7, Server 2012, Windows 10 1809 (on which I was previously able to reproduce #12160 ) and latest preview of Windows 11 started NVDA with both specific language set in preferences and with the language set to default. Made sure that when 'user default' is selected appropriate language is chosen and locale (as retrieved by `locale.setlocale(locale.LC_ALL)`) is set into a locale of a form `englishLanguageName_englishCountryName.ANSICodePage`. Repeated these  tests with some languages known to cause troubles in the past:
- Aragonese - locale set to Windows locale
- German_Switzerland and Hindi - made sure that #12160  has not been reintroduced

Executed unit tests on all above mentioned versions of Windows made sure they pass.
### Known issues with pull request:
None known
### Change log entries:
Bug fixes:
- NVDA no longer sets invalid Python locales

Changes for developers:
- `LOCALE_SLANGUAGE`, `LOCALE_SLIST` and LOCALE_SLANGDISPLAYNAME are moved to the `LOCALE` enum in `languageHandler`. They are still available at the module level but deprecated and would be removed in NVDA 2022.1
### Code Review Checklist:


- [X] Pull Request description is up to date.
- [X] Unit tests.
- [X] System (end to end) tests.
- [X] Manual testing.
- [X] User Documentation.
- [X] Change log entry.
- [X] Context sensitive help for GUI changes.
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

lukaszgo1 committed:
Add various wrappers around `GetLocaleInfoEx` and improve locale normalization for Win32 functions

lukaszgo1 committed:
When setting Python's locale don't rely on `locale.normalize` as it is wrong on Windows rather construct locale string using Win32 functions

lukaszgo1 committed:
Remove work aroud for issue #12160 PR #12250 as it is no longer needed since we're now setting Python's locale to a correct Win32 ones.

lukaszgo1 committed:
Expand / fix unit tests

lukaszgo1 committed:
Improved implementation of `InitLocale` for our instance of `wx.App`

lukaszgo1 committed:
Improve comments - add some docstrings

lukaszgo1 commented:
@seanbudd  Given your work on #12214 you may be interested in taking a look at this.

seanbudd commented on source/languageHandler.py:
```diff
@@ -79,20 +115,90 @@ def getLanguageDescription(language):
+		buf = ctypes.create_unicode_buffer(buffLength)
+		winKernel.kernel32.GetLocaleInfoEx(localeName, LOCALE_SENGLISHLANGUAGENAME, buf, buffLength)
+		langName = buf.value
+		if "Unknown" in langName:
+			# Windows 10 returns 'Unknown' for locales not known to Windows
+			# even though documentation states that in case of an unknown locale 0 is returned.
+			return None
```
Could you add a test to confirm None is returned in this case?

lukaszgo1 replied:
This is being tested by `test_noLangNameFromUnknownLocale`. Note that this test is a general one but that is because it is impossible to know in advance if the version of Windows on which it is executed returns 0 or 'unknown' for unknown locales so it merely checks if `None` is returned for unknown locale. Creating tests which ensures that both these paths results in `None` would be rather difficult, the only way I can thing of at present would be to:
- Create a wrapper around invocation of `GetLocaleInfoEx(localeName, LOCALE_SENGLISHLANGUAGENAME, None, 0)` which returns whatever the underlying system returns without any validation of the result
- Then create two unit tests in which this wrapper is mocked so that it return ''Unknown' for one test and 0 for the second.

I'm not convinced this is worth the trouble.

seanbudd replied:
Thanks, this justification is more than sufficient.

seanbudd commented on source/languageHandler.py:
```diff
@@ -79,20 +115,90 @@ def getLanguageDescription(language):
+				except UnicodeEncodeError:
+					continue
+			return " ".join(partsList)
+	return None
```
Could you add a test to confirm None is returned in this case?

lukaszgo1 replied:
This is being tested by `test_noLangNameFromUnknownLocale` as explained above.

seanbudd commented on source/languageHandler.py:
```diff
@@ -79,20 +115,90 @@ def getLanguageDescription(language):
+	return None
+
+
+def englishCountryNameFromNVDALocale(localeName: str) -> Optional[str]:
```
Could you also add tests for the `None` return values for this function?

lukaszgo1 replied:
This is being tested  by `test_noCountryNameFromUnknownLocale`

seanbudd commented on source/languageHandler.py:
```diff
@@ -23,6 +24,12 @@
+LOCALE_SENGLISHCOUNTRYNAME = 0x00001002
+LOCALE_IDEFAULTANSICODEPAGE = 0x00001004
+
+CP_ACP = "0"
```
This variable name isn't very descriptive, is this an accurate description?

```suggestion
CODEPAGE_ANSICODEPAGE = "0"
```

lukaszgo1 replied:
This is how Microsoft names this constant and for someone familiar with the encoding handling in Windows this is pretty clear. I've hopefully clarified with a comment (853f765a2) let me know if it makes things better.

seanbudd commented on source/languageHandler.py:
```diff
@@ -23,6 +24,12 @@
 LOCALE_SLANGDISPLAYNAME=0x6f
 LOCALE_USER_DEFAULT = 0x400
 LOCALE_CUSTOM_UNSPECIFIED = 0x1000
+LOCALE_SENGLISHLANGUAGENAME = 0x00001001
+LOCALE_SENGLISHCOUNTRYNAME = 0x00001002
+LOCALE_IDEFAULTANSICODEPAGE = 0x00001004
```
Would you be able to turn these new constants into an `IntEnum`?

Considering the amount of work that has gone into this already, and the complexity of this PR, I'm happy to create an issue to move the other related constants into the `IntEnum`, and prepare the deprecations myself.

lukaszgo1 replied:
Done as of d2b973636. The PR description has been updated with an appropriate change log entry for developers. I haven't moved The remaining module level constants such as `LOCALE_CUSTOM_UNSPECIFIED` to the enum since they're not used when retrieving locale info they just  happens to share the same prefix `LOCALE`

seanbudd commented on source/languageHandler.py:
```diff
@@ -79,20 +115,90 @@ def getLanguageDescription(language):
+	return None
+
+
+def ansiCodePageFromNVDALocale(localeName: str) -> Optional[str]:
```
Could you also add tests for the `None` return values for this function?

lukaszgo1 replied:
This is being tested by `test_noCodePageFromUnknownLocale`

seanbudd replied:
my mistake

seanbudd commented on source/languageHandler.py:
```diff
@@ -213,76 +319,92 @@ def setLanguage(lang: str) -> None:
+				log.debug(f"Win32 locale string from locale code is {localeString}")
+			except ValueError:
+				log.debugWarning(f"Locale {localeName} not supported by Windows")
+	if localeString:
```
Could this be changed to:

```suggestion
	if localeString and failedToSetLocale :
```

From what I can tell it does not change the behaviour, and make it more clear at a glance what happens if the `return` statement is missed.

seanbudd commented on source/languageHandler.py:
```diff
@@ -213,76 +319,92 @@ def setLanguage(lang: str) -> None:
-		# Python couldn't support the language_country locale, just try language.
+			failedToSetLocale = True
+			log.debugWarning(f"python locale {localeString} could not be retrieved with getlocale")
+		# The full form langName_country either cannot be retrieved from Windows
+		# or Python cannot be set to that locale.
+		# Try just with the language name.
+		if "_" in localeName:
+			localeName = localeName.split("_")[0]
+			try:
+				localeString = localeStringFromLocaleCode(localeName)
+				log.debug(f"Win32 locale string from locale code is {localeString}")
+			except ValueError:
+				log.debugWarning(f"Locale {localeName} not supported by Windows")
```
I think this section can have the indentation reduced, taking it outside of the above `if localeString:` block.

```suggestion
	# The full form langName_country either cannot be retrieved from Windows
	# or Python cannot be set to that locale.
	# Try just with the language name.
	if "_" in localeName:
		localeName = localeName.split("_")[0]
		try:
			localeString = localeStringFromLocaleCode(localeName)
			log.debug(f"Win32 locale string from locale code is {localeString}")
		except ValueError:
			log.debugWarning(f"Locale {localeName} not supported by Windows")
```

seanbudd commented on source/languageHandler.py:
```diff
@@ -213,76 +319,92 @@ def setLanguage(lang: str) -> None:
-		# as the locale may have been changed to something that getlocale() couldn't retrieve
+			failedToSetLocale = True
+			log.debugWarning(f"python locale {localeString} could not be retrieved with getlocale")
+		# As a final fallback try setting locale just to the English name of the given language.
+		localeFromLang = englishLanguageNameFromNVDALocale(localeName)
+		if localeFromLang:
+			try:
+				locale.setlocale(locale.LC_ALL, localeFromLang)
+				locale.getlocale()
+				log.debug(f"set python locale to {localeFromLang}")
+				return
+			except locale.Error:
+				failedToSetLocale = True
+				log.debugWarning(f"python locale {localeFromLang} could not be set")
+			except ValueError:
+				failedToSetLocale = True
+				log.debugWarning(f"python locale {localeFromLang} could not be retrieved with getlocale")
```
I think the indentation can be reduced here too

```suggestion
	# As a final fallback try setting locale just to the English name of the given language.
	localeFromLang = englishLanguageNameFromNVDALocale(localeName)
	if localeFromLang:
		try:
			locale.setlocale(locale.LC_ALL, localeFromLang)
			locale.getlocale()
			log.debug(f"set python locale to {localeFromLang}")
			return
		except locale.Error:
			failedToSetLocale = True
			log.debugWarning(f"python locale {localeFromLang} could not be set")
		except ValueError:
			failedToSetLocale = True
			log.debugWarning(f"python locale {localeFromLang} could not be retrieved with getlocale")

```

seanbudd reviewed (changes requested):
Thanks @lukaszgo1 for your impressive work on this.

I've left some notes on code styling, and requested some additional tests, but the logic here looks solid.
Once we can get confirmation that this works for users on other systems, as [requested here](https://github.com/nvaccess/nvda/issues/12160#issuecomment-902212282), I think this can be merged. If this can't be done in a reasonable timeframe, we can move forward to alpha testing.

lukaszgo1 committed:
Clarify CP_ACP with a comment

lukaszgo1 committed:
Move all NLS constants into an enum

lukaszgo1 committed:
Simplify implementation of `setLocale`

lukaszgo1 committed:
Revert "Remove work aroud for issue #12160 PR #12250 as it is no longer needed since we're now setting Python's locale to a correct Win32 ones."

This reverts commit d2491a4a150e83c950b5e13dab3c4e4a5cb1b224.

lukaszgo1 committed:
Clarifying comments

lukaszgo1 commented:
@seanbudd  I believe all your review comments are now addressed. I've simplified the implementation of `setLocale` though had to do so differently than you've suggested since after applying your suggestions Linter started to complain about  it being too complex. Contrary to my initial belief I also had to keep the code from #12250 - the updated PR description explains in which cases it is necessary. The flip side of this is that it is now much safer to merge this into Alpha.

seanbudd commented on source/languageHandler.py:
```diff
@@ -213,85 +337,88 @@ def setLanguage(lang: str) -> None:
+		log.debugWarning(f"Locale {localeName} not supported by Windows")
+	if localeString and _setPythonLocale(localeString):
+		return
+	else:
+		# The full form langName_country either cannot be retrieved from Windows
+		# or Python cannot be set to that locale.
+		# Try just with the language name.
+		if "_" in localeName:
+			localeName = localeName.split("_")[0]
+			try:
+				localeString = localeStringFromLocaleCode(localeName)
+				log.debug(f"Win32 locale string from locale code is {localeString}")
+			except ValueError:
+				log.debugWarning(f"Locale {localeName} not supported by Windows")
```
This could also be simplified

```suggestion
	# The full form langName_country either cannot be retrieved from Windows
	# or Python cannot be set to that locale.
	# Try just with the language name.
	if "_" in localeName:
		localeName = localeName.split("_")[0]
		try:
			localeString = localeStringFromLocaleCode(localeName)
			log.debug(f"Win32 locale string from locale code is {localeString}")
		except ValueError:
			log.debugWarning(f"Locale {localeName} not supported by Windows")
```

seanbudd commented on source/languageHandler.py:
```diff
@@ -213,85 +337,88 @@ def setLanguage(lang: str) -> None:
+				log.debugWarning(f"Locale {localeName} not supported by Windows")
+	if localeString and _setPythonLocale(localeString):
+		return
+	else:
+		# As a final fallback try setting locale just to the English name of the given language.
+		localeFromLang = englishLanguageNameFromNVDALocale(localeName)
+		if localeFromLang and _setPythonLocale(localeFromLang):
+			return
```
Same here

```suggestion
	# As a final fallback try setting locale just to the English name of the given language.
	localeFromLang = englishLanguageNameFromNVDALocale(localeName)
	if localeFromLang and _setPythonLocale(localeFromLang):
		return
```

seanbudd commented on source/languageHandler.py:
```diff
@@ -213,85 +337,88 @@ def setLanguage(lang: str) -> None:
-			log.debugWarning(f"setting python locale to the current language {curLang}")
-			# fallback and try to reset the locale to the current lang
-			setLocale(curLang)
+			# Either Windows does not know the locale, or Python is unable to handle it.
+			# reset to default locale
+			if originalLocaleName == curLang:
+				# reset to system locale default if we can't set the current lang's locale
+				locale.setlocale(locale.LC_ALL, "")
+				log.debugWarning(f"set python locale to system default")
+			else:
+				log.debugWarning(f"setting python locale to the current language {curLang}")
+				# fallback and try to reset the locale to the current lang
+				setLocale(curLang)
```
I would remove the above `else` and also de-indent here

```suggestion
	# Either Windows does not know the locale, or Python is unable to handle it.
	# reset to default locale
	if originalLocaleName == curLang:
		# reset to system locale default if we can't set the current lang's locale
		locale.setlocale(locale.LC_ALL, "")
		log.debugWarning(f"set python locale to system default")
	else:
		log.debugWarning(f"setting python locale to the current language {curLang}")
		# fallback and try to reset the locale to the current lang
		setLocale(curLang)
```

seanbudd commented on source/logHandler.py:
```diff
@@ -290,7 +290,8 @@ def formatException(self, ex):
 		"""Custom implementation of `formatTime` which avoids `time.localtime`
-		since it causes a crash under some versions of Universal CRT ( #12160, Python issue 36792)
+		since it causes a crash under some versions of Universal CRT when Python locale
+		is set to a Unicode one ( #12160, Python issue 36792)
```
Might as well fix this spacing while here

```suggestion
		is set to a Unicode one (#12160, Python issue 36792)
```

seanbudd reviewed (commented):
Thanks for implementing the enum and deprecations and pointing out the tests I missed. This looks almost ready to merge.

I've requested some further changes to `setLocale` to make it easier to visually parse. I'm not sure if this sort of code-styling is helpful for vision impaired developers - if it is detrimental I am happy to ignore it.

lukaszgo1 committed:
Merge branch 'master' into correctWinLocales

lukaszgo1 committed:
Review actions

lukaszgo1 commented:
@seanbudd  All addressed.

seanbudd reviewed (approved):
Thanks @lukaszgo1, this change will be much appreciated

seanbudd committed:
Merge branch 'master' into correctWinLocales

seanbudd committed:
update changes
